Feat/#114 add asset UI

superozing opened the pull request:
![스크린샷 2024-05-16 230244](https://github.com/Remnants-of-Remnants/DirectX3D/assets/127123002/cfd4c841-4d8b-4f50-93a8-956090a3e479)
![스크린샷 2024-05-16 230434](https://github.com/Remnants-of-Remnants/DirectX3D/assets/127123002/fd6131f7-0563-4251-93c2-fdee41695e01)

superozing committed:
#114 Refactor / Engine Asset도 material의 이름이 뜨도록 수정

Engine Asset은 파일 경로가 없기 때문에 이름이 설정되지 않았는데,
키 값을 가져와서 Engine Asset도 이름을 표시하도록 수정했어요.

superozing committed:
#114 Feat / Graphics Shader가 Shader 정보를 가지도록 멤버 추가

CreateShader() 함수를 호출하고 Shader 생성에 성공한 시점에 인자로 받아온 문자열을 멤버로 저장해요.

Shader 문자열 정보와 State 정보를 반환하는 Get함수를 추가했어요.

superozing committed:
#114 Feat / GraphicsShaderUI에서 Shader 정보 표시

경로와 함수 명을 표시하고, 각각의 state를 표시해요.

superozing committed:
#114 Feat / enum string 멤버 추가

매번 매직 이넘을 사용해서 가져오는 것 보다는, 초기화 시점에 한 번 설정해주고 매 번 가져오는 것이 좋아보여서
ImGuiMgr의 멤버로 enum값에 대응하는 string을 들고 있도록 해주었어요.

superozing committed:
#114 Feat / Add GraphicsShader State Combo Box

매직 이넘은 신이야

superozing committed:
#114 Refactor / scalar, tex param을 배열로 교체

superozing committed:
#114 Feat / IsUse 추가 -> 해당 param이 사용되는지 체크

superozing committed:
#114 Feat / 동적으로 param 사용 여부와 desc 수정 기능 추가

superozing commented:
![스크린샷 2024-05-21 175855](https://github.com/Remnants-of-Remnants/DirectX3D/assets/127123002/088af9ca-878b-4808-a506-7e1b379a6a3d)

![스크린샷 2024-05-21 175906](https://github.com/Remnants-of-Remnants/DirectX3D/assets/127123002/69772bde-e426-422c-843f-0706ef27ca56)

![스크린샷 2024-05-21 175931](https://github.com/Remnants-of-Remnants/DirectX3D/assets/127123002/a0d4e987-24da-4a66-9282-4ffb62d9b528)

zkzkzkzkzkzz reviewed (commented):
동적으로 바꿨을 때 바꾼 설정이 저장이 되는건가?

superozing commented:
저장이 안되고 있는데... 어떤 식으로 해야 할 지 감이 안잡히네

superozing committed:
#114 Feat / GraphicsShader Save(), Load() 추가

superozing committed:
#114 Bug / param을 올바른 곳에 넣어주기

jeontigger committed:
#114 Feat / 그래픽스 셰이더 파일화, 세이브 로드 구현

jeontigger committed:
#114 Feat / 그래픽스 셰이더 파일 코드제네레이터 읽기, 예외 처리 메시지 반복 안하도록 처리

superozing committed:
#114 Feat / Load 한 GraphicsShader 정보를 표시

폴더 명을 임시로 없애주었어요.

머터리얼도 파일화가 필요해 보인다.

superozing committed:
#114 Feat / 레벨 전환 시 GraphicsShader를 삭제하지 않도록 수정

이제 그래픽스 쉐이더를 파일로부터 불러오기 때문에 그래픽스 쉐이더는 엔진 에셋이 아니에요.

그래픽스 쉐이더가 삭제 될 경우 기존 쉐이더가 완전히 삭제되지 않고 참조 개수가 남아 머터리얼이 그대로 들고 있기 때문에, 그래픽스 쉐이더는 레벨이 전환되어도 삭제하지 않도록 해주었어요.

superozing committed:
#114 Refactor / Graphics Shader는 반드시 엔진 에셋으로 설정

레벨 전환 시 지워주지 않기 위해서.

superozing committed:
#114 Refactor / GraphicsShader 생성자 기본 인자 제거

반드시 엔진 에셋으로 생성하기 위해서 에요.

superozing committed:
#114 Feat / filesystem 활용한 파일 이름 가져오기

superozing committed:
#114 Feat / PopupModal 이용한 GraphicsShader 생성, 저장

그래픽스 쉐이더를 생성하는 버튼을 누를 경우 팝업 창을 이용해 쉐이더 파일 상대 경로와 함수 명을 받아 컴파일 해요.
(함수 명과 파일 명을 set 하면서, 함수 명이나 파일 명이 잘못되었을 경우의 예외 처리도 겸사겸사 하기 위해서.)

superozing commented:
![스크린샷 2024-05-28 181114](https://github.com/Remnants-of-Remnants/DirectX3D/assets/127123002/25da8c68-07fc-4e37-9743-96e03674981c)

superozing committed:
#114 Feat / 원본 GraphicsShader 생성 버튼

superozing committed:
#114 Feat / Add GraphicsShader Save Button

oljindudol committed:
Merge pull request #155 from Remnants-of-Remnants/Feat/#8-FollowUp-240531-FRI

Feat/#8 follow up 240531 fri

superozing committed:
Merge branch 'dev' into Feat/#114-Add-AssetUI

superozing committed:
Merge branch 'dev' of https://github.com/Remnants-of-Remnants/DirectX3D into dev

superozing committed:
Merge branch 'dev' into Feat/#114-Add-AssetUI
